### ISSUE TITLE
[require-atomic-updates]: False positive on reused loop variable name

### ISSUE THREAD

**Issue by fluggo** [eslint/eslint#14208]
**Tell us about your environment**

Environment Info:

Node version: v14.15.4
npm version: v7.5.3
Local ESLint version: v7.21.0 (Currently used)
Global ESLint version: Not found

Parser: `@typescript-eslint/parser`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  env: {
    es2020: true,
    node: true
  },
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    'project': './tsconfig.eslint.json',
    'sourceType': 'module',
    'tsconfigRootDir': __dirname,
  },
  plugins: [
    'eslint-plugin-import',
    '@typescript-eslint',
  ],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-requiring-type-checking',
  ],
  rules: {
    '@typescript-eslint/no-non-null-assertion': 'warn',
    '@typescript-eslint/no-shadow': ['error', { 'hoist': 'all' }],
    '@typescript-eslint/no-unused-expressions': 'error',
    '@typescript-eslint/require-await': 'warn',
    '@typescript-eslint/require-array-sort-compare': 'error',
    '@typescript-eslint/no-base-to-string': 'error',
    '@typescript-eslint/restrict-plus-operands': 'error',
    'array-callback-return': 'error',
    'eqeqeq': ['error', 'always'],
    'guard-for-in': 'error',
    'no-bitwise': 'error',
    'no-caller': 'error',
    'no-eval': 'error',
    'no-new-wrappers': 'error',
    'no-template-curly-in-string': 'error',
    'no-throw-literal': 'error',
    'no-undef-init': 'error',
    'no-unreachable-loop': 'error',
    'no-useless-backreference': 'error',
    'no-var': 'error',
    'require-atomic-updates': 'error',
    'prefer-const': 'off',      // change to 'error'
    '@typescript-eslint/consistent-type-definitions': ['warn', 'interface'],
    '@typescript-eslint/semi': ['error', 'always'],
    '@typescript-eslint/unified-signatures': 'error',
    '@typescript-eslint/prefer-string-starts-ends-with': 'warn',
    '@typescript-eslint/no-extra-semi': 'off',        // change to 'warn'
    '@typescript-eslint/no-inferrable-types': 'off',    // change to ['warn', {ignoreParameters: true}]
    'arrow-body-style': 'off',          // change to ['error', 'as-needed'],
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    '@typescript-eslint/no-unused-vars': 'off',
    '@typescript-eslint/no-empty-function': 'off',
    '@typescript-eslint/no-unsafe-assignment': 'off',
    '@typescript-eslint/no-unsafe-member-access': 'off',
    '@typescript-eslint/no-unsafe-call': 'off',
    '@typescript-eslint/no-unsafe-return': 'off',
    '@typescript-eslint/no-implicit-any-catch': 'off',
    '@typescript-eslint/restrict-template-expressions': 'off',
    '@typescript-eslint/no-unnecessary-condition': 'off',
    'no-inner-declarations': 'off',
    'no-promise-executor-return': 'off',
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
async function foo(e) {
}

async function run() {
  const input = [];
  const props = [];

  for(const entry of input) {
    const prop = props.find(a => a.id === entry.id) || null;
    await foo(entry);
  }

  // Possible race condition: `const entry` might be reassigned based on an outdated value of `const entry`
  for(const entry of input) {
    const prop = props.find(a => a.id === entry.id) || null;
  }

  for(const entry2 of input) {
    const prop = props.find(a => a.id === entry2.id) || null;
  }
}
```

Command: `npx eslint scripts/sample-require-atomic-updates.ts`

**What did you expect to happen?**

Did not expect any errors because the `entry` variables in the two loops are independent.

**What actually happened? Please include the actual, raw output from ESLint.**

ESLint seems to think that the variable captured in the second loop's arrow function closure is the `entry` variable from the first loop, and flags an error.

```
/......./scripts/sample-require-atomic-updates.ts
  14:3  error  Possible race condition: `const entry` might be reassigned based on an outdated value of `const entry`  require-atomic-updates

✖ 1 problem (1 error, 0 warnings)
```

**Comment by mdjermanovic** [comment]
Hi @fluggo, thanks for the issue!

I can reproduce this is our [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHJlcXVpcmUtYXRvbWljLXVwZGF0ZXM6IGVycm9yICovXG5cbmFzeW5jIGZ1bmN0aW9uIGZvbyhlKSB7XG59XG5cbmFzeW5jIGZ1bmN0aW9uIHJ1bigpIHtcbiAgY29uc3QgaW5wdXQgPSBbXTtcbiAgY29uc3QgcHJvcHMgPSBbXTtcblxuICBmb3IoY29uc3QgZW50cnkgb2YgaW5wdXQpIHtcbiAgICBjb25zdCBwcm9wID0gcHJvcHMuZmluZChhID0+IGEuaWQgPT09IGVudHJ5LmlkKSB8fCBudWxsO1xuICAgIGF3YWl0IGZvbyhlbnRyeSk7XG4gIH1cblxuICAvLyBQb3NzaWJsZSByYWNlIGNvbmRpdGlvbjogYGNvbnN0IGVudHJ5YCBtaWdodCBiZSByZWFzc2lnbmVkIGJhc2VkIG9uIGFuIG91dGRhdGVkIHZhbHVlIG9mIGBjb25zdCBlbnRyeWBcbiAgZm9yKGNvbnN0IGVudHJ5IG9mIGlucHV0KSB7XG4gICAgY29uc3QgcHJvcCA9IHByb3BzLmZpbmQoYSA9PiBhLmlkID09PSBlbnRyeS5pZCkgfHwgbnVsbDtcbiAgfVxuXG4gIGZvcihjb25zdCBlbnRyeTIgb2YgaW5wdXQpIHtcbiAgICBjb25zdCBwcm9wID0gcHJvcHMuZmluZChhID0+IGEuaWQgPT09IGVudHJ5Mi5pZCkgfHwgbnVsbDtcbiAgfVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=), looks like a bug..

**Comment by patriscus** [comment]
I am currently working on a potential fix. I hope to create a PR this weekend
